docs/env-config.md
# Fin Configuration 

The following are the standard environmental variables used by Fin and it's core services. Developers, all config properties can be seen here: [config.js](../services/fin/node-utils/config.js) 

  - [Base Config](#base-config)
  - [Authentication](#authentication)
  - [Google Cloud Config](#google-cloud-config)
  - [Postgres Config](#postgres-config)
  - [ElasticSearch Config](#elasticsearch-config)
  - [Redis Config](#redis-config)

# Base Config

## FIN_URL (required)

Default: http://localhost:3000

This should be the root url (domain name) of your Fin service. 

## GCS_BUCKET_ENV

Default: local-dev

Used for accessing Google Cloud Storage data environments.  Example in the GCS config definitions you can
say `dams-client-{{GCS_BUCKET_ENV}}`, which will resolve to `dams-client-local-dev` or `dams-client-prod` depending on the value of `GCS_BUCKET_ENV` and which environment you are running in.

## WORKFLOW_ENV

Default: local-dev

Used for accessing Google Cloud Workflow environments.  Example in the GCS config definitions you can
say `dams-client-{{GCS_BUCKET_ENV}}`, which will resolve to `dams-client-local-dev` or `dams-client-prod` depending on the value of `GCS_BUCKET_ENV` and which environment you are running in.

## CLIENT_ENV

Default: dev

Should be used be your client application.  Normally `prod` serves your production 
build of the client app.  Everything else serves the dev build.  But this is not
implemented by fin, you must implement it in your client app.

## LOG_LEVEL FIN_LOG_LEVEL

Default: info

Used to set the log level for all services.  Either env var will work.

## FIN_MODEL_ROOT

Default: /fin/services/models

Path to the directory where your fin data models are stored.


## FIN_COOKIE_SECRET

Default: changeme

Used to encrypt your cookies

## FIN_COOKIE_MAX_AGE (time in milliseconds)

Default: 7 days

## FIN_ALLOW_ORIGINS

Comma separated list of origins you would like to grant access to FIN.  Requests from these origins will set proper CORS headers in the response as well as handle the browser preflight OPTIONS CORS request.

[Read more](./cors.md)

# Authentication

## JWT_SECRET (required)

Secret used for talking to OIDC provider (keycloak)

## JWT_ISSUER (required)

Issuer used for talking to OIDC provider (keycloak)

## JWT_COOKIE_NAME

Default: fin-jwt

Name of cookie to store JWT token

## JWT_JWKS_URI (required)

URL to the JWKS endpoint of the OIDC provider (keycloak)

## OIDC_CLIENT_ID (required)

Client ID of the OIDC provider client (keycloak)

## OIDC_CLIENT_SECRET (required)

Client secret of the OIDC provider client (keycloak)

## OIDC_BASE_URL (required)

Base URL of the OIDC provider realm (keycloak)

## OIDC_SCOPES

Default: roles openid profile email

Scopes to request in token

## OIDC_FIN_LDP_SERVICE_NAME

Default: keycloak-oidc

Name of the OIDC auth services (so container name in `docker compose` speak) that is used to authenticate against keycloak.

## FIN_SERVICE_ACCOUNT_NAME

Service account name used to authenticate to the OIDC provider (keycloak). For keycloak, this is the username of the service account.

## FIN_SERVICE_ACCOUNT_PASSWORD

Service account password used to authenticate to the OIDC provider (keycloak). For keycloak, this is the password of the service account.  This should be a 512 character string.

```bash
openssl rand -base64 512
```

If you have the above `FIN_SERVICE_ACCOUNT_*` parameters set in your `.env`.  You can test your service account with:

```bash
docker run --rm -i -t --env-file .env --name init us-west1-docker.pkg.dev/digital-ucdavis-edu/pub/fin-init:${version} node /service/getToken.js
```

# Google Cloud Config

## GOOGLE_APPLICATION_CREDENTIALS

Default: /etc/fin/service-account.json

Path to the service account json file.  This is used by any service to authenticate to GCS.

## GOOGLE_SERVICE_ACCOUNT_EMAIL

Service account to use for GCS.  If not set, the service account specified in the GOOGLE_APPLICATION_CREDENTIALS file will be used.

## GOOGLE_CLOUD_PROJECT

Google Cloud Project ID.  If not set, the project ID specified in the GOOGLE_APPLICATION_CREDENTIALS file will be used.

## GOOGLE_CLOUD_LOCATION

Default: us-central1

Location of GCS services.

## GOOGLE_PUBSUB_SUBSCRIPTION_NAME

Name to use for pubsub subscriptions.  If not set, the default subscription name will be set to `GCS_BUCKET_ENV`, which defaults to
`local-dev` if not set. 

## GOOGLE_MAX_CONCURRENT_WORKFLOWS

Default: 3

Set the maximum number of concurrent workflows to run.  This is used by the workflow service to limit the number of workflows that can be running at once.  This only applies to workflows of type `gc-workflow`.

# Postgres Config

## PG_HOST

Default: postgres

## PG_PORT

Default: 5432

## PG_USER

Default: postgres

## PG_DATABASE

Default: fcrepo


# ElasticSearch Config

## ES_HOST

Default: elasticsearch

## ES_PORT

Default: 9200

## ELASTIC_USERNAME

Default: elastic

## ELASTIC_PASSWORD

Default: elastic

## ES_LOG_LEVEL

Default: error

# Redis Config

## REDIS_HOST

Default: redis

## REDIS_PORT

Default: 6379

# Fcrepo Config

## FCREPO_HOST

Default: fcrepo

Host name of the `fcrepo` service


CHANGELOG.md
# 2.7.0

  - gateway moved to port 3000

# 2.5.0

  - Validation updates (major db schema change)
  - Admin UI updates

# 2.3.1
  
  - doc updates
  - added wrapper method to activemq tests to simplify ping 
  - added logo to admin console

# 2.3.0
  
  - Switched to RabbitMQ for messaging
  - Added AMPQ jars to fcrepo
  - Changed messaging format to be jsonld and no longer using headers.
# 2.2.7

  - admin ui updates
  - fin-cache updates.  quads_view requires manual update

# 2.2.6

  - gcssync bug fixes
  - direct ocfl access bug fixes
  - fin-cache fixes

# 2.2.5

  - gcssync bug fixes and updates

# 2.2.4

  - fin io calculating sha256, sha512 and md5 checksums for binary files
  - fin quad cache now storing last modified date for quads
  - fin direct ocfl access to fetching last modifed and created dates
  - major bug fixes for gcssync gcs -> fcrepo
  - fin io adding md5 and sha512 checksums for metadata files

# 2.2.2

  - fin io no longer looks at `@id` for ldp container id lookup
  - Admin UI displays changes for (some) values, this will by built out more
  - Data validation entries are now deleted if not database item exists
  - dbsync.update_status entries for update_types `Delete` and `Purge` are now deleted removed after one week.

# 2.2.1

  - fixes for fin io `@base:` handling
  - fix for dbsync reindex endpoint
  - fix for stomp disconnection
  - adding force binary update flag to fin io
  - fix for acl imports via fin io (@base: required for pointing at container)

# 2.2.0

  - `fin io` indirect reference restructure
 
# 2.1.x

  - fin io import rewrite.
  - gcssync fixes

# 2.0.4

 - Fix for gcssync when root path is a `gcs` 'directory' (i.e. a real file)
 - GCS sync now creates sub paths if they don't exist ensure that they can by modified later on

# 2.0.3

 - Disabling autoversioning of ocfl.  `fcrepo.autoversioning.enabled=false`
 - Fix for exposing pg uuid to public schema
 - update for data model fcrepo access when no transform provided
 - Start of new `/fin/rest` endpoint for; direct access to OCFL via fs and pg, always returns expanded JSONLD even for binaries, so no need for `/fcr:metadata`.

# 2.0.2
 
 - Fix For `fin io` command where `iodir` was just checking the graph context

# 2.0.1

 - First versioned image release of Fin v2


services/fin/node-utils/config.js
const fs = require('fs');
const path = require('path');

// load custom .env file if it exists in environment variables
// this is really useful for k8s environments where individual 
// env variables from secrets are verbose and hard to manage
let envPath = '/etc/fin/.env';
if( process.env.FIN_ENV_FILE ) {
  envPath = process.env.FIN_ENV_FILE;
}
if( fs.existsSync(envPath) && fs.lstatSync(envPath).isFile() ) {
  if( ['info', 'debug'].includes(process.env.LOG_LEVEL) ) {
    console.log(`Loading environment variables from ${envPath}`);
  }
  require('dotenv').config({ path: envPath });
}

const env = process.env;
const COMMON_URI = require('./lib/common-rdf-uris');

function processArray(value) {
  if( value && value.length ) {
    return value.split(/,|\s/)
      .map(item => item.trim())
      .filter(item => item);
  }
  return null;
}

var fcrepoHostname = process.env.FCREPO_HOST || 'fcrepo';
var fcrepoPort = process.env.FCREPO_PORT || '8080';
if( fcrepoPort.match(/^tcp:\//) ) {
  fcrepoPort = new URL(fcrepoPort).port;
}

var esHostname = process.env.ES_HOST || 'elasticsearch';
var esPort = process.env.ES_PORT || 9200;

let serviceAccountFile = process.env.GOOGLE_APPLICATION_CREDENTIALS || '/etc/fin/service-account.json';
let serviceAccountExists = fs.existsSync(serviceAccountFile) && fs.lstatSync(serviceAccountFile).isFile();
let gcServiceAccount = {};
if( serviceAccountExists ) {
  if( !env.GOOGLE_APPLICATION_CREDENTIALS ) {
    env.GOOGLE_APPLICATION_CREDENTIALS = serviceAccountFile;
  }
  gcServiceAccount = JSON.parse(fs.readFileSync(serviceAccountFile, 'utf-8'));
}

let dbsyncIgnoreEnv = processArray(process.env.DBSYNC_IGNORE_TYPES);

let esCompactTypesInclude = processArray(process.env.ES_COMPACT_TYPES_INCLUDE);
if( esCompactTypesInclude ) {
  esCompactTypesInclude = esCompactTypesInclude.map(type => new RegExp(type));
}

let esFieldsExclude = processArray(process.env.ES_FIELDS_EXCLUDE);
let esFieldsExcludeCompact = processArray(process.env.ES_FIELDS_EXCLUDE_COMPACT);
let gcsDiskCacheExts = processArray(process.env.GCS_DISK_CACHE_EXTS);
let disableServices = processArray(process.env.DISABLE_FIN_SERVICES);

// make sure this is set for gcssync templates
if( !env.WORKFLOW_ENV ) env.WORKFLOW_ENV = 'local-dev';

let finCachePredicates = (processArray(process.env.FIN_CACHE_PREDICATES) || [])
  .map(predicate => {
    if( predicate.match(/^\/.*\/(i|g)?$/) ) {
      let flags = predicate.split('/').pop();
      return new RegExp(predicate.replace(/(^\/|\/(g|i)?$)/g, ''), flags);
    } else {
      return predicate;
    }
  });
if( finCachePredicates.length === 0 ) {
  finCachePredicates = [
    new RegExp('^http://digital.ucdavis.edu/schema#')
    // currently not used as hasMessageDigest is in the digital namespace via the fin-digests service
    // 'http://www.loc.gov/premis/rdf/v1#hasMessageDigest'
  ];
}

let extensionRoutes = [];
if( env.ADMIN_UI_EXTENSIONS_ROUTES ) {
  extensionRoutes = processArray(env.ADMIN_UI_EXTENSIONS_ROUTES).map(eleRoute => {
    let [path, element] = eleRoute.split(':');
    return {path, element};
  });
}


module.exports = {

  projectName : env.PROJECT_NAME || 'fin',
  serviceName : env.FIN_SERVICE_NAME || 'unknown',

  serviceAccount : {
    username : env.FIN_SERVICE_ACCOUNT_NAME,
    secret : env.FIN_SERVICE_ACCOUNT_SECRET
  },

  server : {
    url : process.env.FIN_URL || 'http://localhost:3000',
    loglevel : process.env.LOG_LEVEL || process.env.FIN_LOG_LEVEL || 'info',
    cookieSecret : process.env.FIN_COOKIE_SECRET || 'changeme',
    cookieMaxAge : process.env.FIN_COOKIE_MAX_AGE ? parseInt(process.env.SERVER_COOKIE_MAX_AGE) : (1000 * 60 * 60 * 24 * 7),
    allowOrigins : (process.env.FIN_ALLOW_ORIGINS || '').split(',').filter(domain => domain !== '').map(domain => domain.trim())
  },

  gateway : {
    http : {
      port : env.FIN_GATEWAY_HTTP_PORT || 3000
    },
    https : {
      enabled : env.FIN_GATEWAY_HTTPS_ENABLED === 'true',
      certFolder : env.FIN_GATEWAY_HTTPS_CERT_FOLDER || '/etc/fin/certs',
      port : env.FIN_GATEWAY_HTTPS_PORT || 3443
    },
    proxy : {
      timeout : env.FIN_GATEWAY_TIMEOUT || 1000 * 60 * 5,
      proxyTimeout : env.FIN_GATEWAY_PROXY_TIMEOUT || 1000 * 60 * 5
    },
    host : 'http://gateway:'+(env.FIN_GATEWAY_HTTP_PORT || 3000),
    fcrepoDataMount : env.GATEWAY_FCREPO_DATA_MOUNT || '/data',
    ocflRoot : 'ocfl-root',
    disableServices
  },

  ocfl : {
    root : env.OCFL_ROOT || '/data/ocfl-root',
    mutableHead : (env.OCFL_MUTABLE_HEAD === 'true' || env.FCREPO_AUTO_VERSION === 'false'),
    directAccess : {
      aclCacheExpire : 1000 * 10
    }
  },

  rdf : {
    baseUrl : 'http://digital.ucdavis.edu/schema#',
    prefix : 'ucdlib'
  },

  fcrepo : {
    hostname : fcrepoHostname,
    port : fcrepoPort,
    host : `http://${fcrepoHostname}:${fcrepoPort}`,
    root : '/fcrepo/rest',
    admin : {
      username : env.FCREPO_ADMIN_USERNAME || 'fedoraAdmin',
      password : env.FCREPO_ADMIN_PASSWORD || 'fedoraAdmin'
    },
    roInstances : {
      enabled : env.FCREPO_RO_ENABLED === 'true',
      host : env.FCREPO_RO_HOSTNAME || 'fcrepo-ro',
      port : env.FCREPO_RO_PORT || 8080,
      numInstances : parseInt(env.FCREPO_RO_NUM_INSTANCES || 2)
    },
  },

  metrics : {
    enabled : env.FIN_METRICS_ENABLED === 'true',
    harvestInterval : env.FIN_METRICS_HARVEST_INTERVAL ? parseInt(env.FIN_METRICS_HARVEST_INTERVAL) : (1000 * 60),
    instruments : {
      fs : {
        enabled : env.FIN_METRICS_FS_ENABLED === 'true',
        basePath : env.FIN_METRICS_FS_PATH || '/fs-tests',
        fileSize : env.FIN_METRICS_FS_FILE_SIZE ? parseInt(env.FIN_METRICS_FS_FILE_SIZE) : (1024*128)
      },
      fcrepo : {
        enabled : env.FIN_METRICS_FCREPO_ENABLED === 'true'
      },
      fin : {
        enabled : env.FIN_METRICS_FIN_ENABLED === 'true'
      }
    },
    export : {
      gc : env.FIN_METRICS_EXPORT_GC === 'true',
      stdout : env.FIN_METRICS_EXPORT_STDOUT === 'true'
    }
  },

  activeMq : {
    hostname : fcrepoHostname,
    fcrepoTopic : '/topic/fedora',
    // set to -1 to disable
    testInterval : env.ACTIVEMQ_TEST_INTERVAL ? parseInt(env.ACTIVEMQ_TEST_INTERVAL) : 1000 * 60 * 5,
    fcrepoTestPath : '/activemq',
    queues : {
      dbsync : '/queue/dbsync',
      gcssync : '/queue/gcssync',
    },
    stomp : {
      port : env.STOMP_PORT || 61613,
      maxLineLength : env.STOMP_MAX_LINE_LENGTH || 1024*32,
    },
    mqtt : {
      port : env.MQTT_PORT || 1883,
      fcrepoTopic : 'fedora',
      queues : {
        dbsync : 'dbsync',
        gcssync : 'gcssync',
      },
    },
    amqp : {
      port : env.FCREPO_AMQP_PORT || 5672
    }
  },

  rabbitmq : {
    host : env.RABBITMQ_HOST || 'rabbitmq',
    port : env.RABBITMQ_PORT || 5672,
    adminPort : env.RABBITMQ_ADMIN_PORT || 15672,
    username : env.RABBITMQ_USERNAME || 'guest',
    password : env.RABBITMQ_PASSWORD || 'guest',
    vhost : env.RABBITMQ_VHOST || '/',
    prefetch : env.RABBITMQ_PREFETCH || 1,
    queues : {
      dbsync : 'dbsync',
      gcssync : 'gcssync',
    },
    shovel : {
      name : 'fcrepo-shovel',
      exchange : env.RABBITMQ_MAIN_EXCHANGE || 'fin',
      finActiveMqQueue : 'fin', // do not include /queue
    }
  },

  pg : {
    host : env.PG_HOST || 'postgres',
    port : env.PG_PORT || 5432,
    user : env.PG_USER || 'postgres',
    database : env.PG_DATABASE || 'fcrepo',
    searchPath : ['public', 'dbsync', 'label_service', 'finac']
  },

  jwt : {
    jwksUri : process.env.JWT_JWKS_URI,
    secret : process.env.JWT_SECRET,
    issuer : process.env.JWT_ISSUER,
    // expires in seconds
    ttl : process.env.JWT_TTL ? parseInt(process.env.JWT_TTL) : (60 * 60 * 24 * 14),
    cookieName : process.env.JWT_COOKIE_NAME || 'fin-jwt'
  },

  principal: {
    cookieName : process.env.PRINCIPAL_COOKIE_NAME || process.env.PRINCIPAL_HEADER_NAME || 'fin-principal',
    headerName : process.env.PRINCIPAL_HEADER_NAME || process.env.PRINCIPAL_COOKIE_NAME || 'fin-principal',
    addDomain : process.env.PRINCIPAL_ADD_DOMAIN || null
  },

  api : {
    port : env.FIN_API_PORT || 3004
  },

  oidc : {
    port : env.OIDC_PORT || 3003,
    clientId : env.OIDC_CLIENT_ID,
    baseUrl : env.OIDC_BASE_URL,
    secret : env.OIDC_SECRET,
    scopes : env.OIDC_SCOPES || 'roles openid profile email',
    finLdpServiceName : env.OIDC_FIN_LDP_SERVICE_NAME || 'keycloak-oidc',
    roleIgnoreList : [
      "default-roles-dams",
      "uma_authorization",
      "manage-account",
      "manage-account-links",
      "view-profile",
      "offline_access"
    ],
    // default cache all tokens for 30 seconds before requesting verification again
    tokenCacheTTL : env.OIDC_TOKEN_CACHE_TTL ? parseInt(env.OIDC_TOKEN_CACHE_TTL) : (1000*60*5)
  },

  finCache : {
    predicates : finCachePredicates
  },

  finDigests : {
    basePath : env.FIN_DIGESTS_BASE_PATH || '/fin/digests'
  },

  finac : {
    port : env.FINAC_PORT || 3002,
    agents : {
      admin : 'admin',
      discover : 'discover',
      protected : 'protected',
      public : 'public'
    },
    defaultAccessTime : 60 * 60 * 3 // 3 hours
  },

  dbsync : {
    ignoreTypes : dbsyncIgnoreEnv || [
      COMMON_URI.TYPES.FIN_IO_INDIRECT,
      COMMON_URI.TYPES.WEBAC
    ]
  },

  elasticsearch : {
    host : esHostname,
    port : esPort,
    username : process.env.ELASTIC_USERNAME || 'elastic',
    password : process.env.ELASTIC_PASSWORD || 'elastic',
    get connStr () {
      return `http://${this.host}:${this.port}`
    },
    log : process.env.ES_LOG_LEVEL || 'error',
    compactTypeInclude : esCompactTypesInclude || [
      new RegExp('http://digital.ucdavis.edu/schema#'),
      new RegExp('http://schema.org/')
    ],
    fields : {
      exclude : esFieldsExclude || [
        'roles',
        '@graph.indexableContent',
        '@graph.createdBy', '@graph.lastModifiedBy', '@graph._', '@graph.textIndexable'
      ],
      excludeCompact : esFieldsExcludeCompact || [
        'roles',
        '@graph.indexableContent',
        '@graph.createdBy', '@graph.lastModifiedBy', '@graph._',
        '@graph.image', '@graph.textIndexable', '@graph.lastModified'
      ]
    }
  },

  esIndexManagement : {
    port : env.ES_INDEX_MANAGEMENT_PORT || 3001,
  },

  redis : {
    host : process.env.REDIS_HOST || 'redis',
    port : process.env.REDIS_PORT || 6379,
    refreshTokenExpire : (86400 * 30)
  },

  models : {
    rootDir : env.FIN_MODEL_ROOT || '/fin/services/models'
  },

  workflow : {
    finConfigPath : '/fin/workflows/config.json',
  },

  adminUi : {
    extensions : {
      enabled : env.ADMIN_UI_EXTENSIONS_ENABLED === 'true',
      sourcePath : env.ADMIN_UI_EXTENSIONS_SOURCE_PATH || '/js/app-extension.js',
      routes : extensionRoutes
    }
  },

  google : {
    serviceAccountExists,
    serviceAccountFile,
    serviceAcountEmail : env.GOOGLE_SERVICE_ACCOUNT_EMAIL || gcServiceAccount.client_email,
    project : env.GOOGLE_CLOUD_PROJECT || gcServiceAccount.project_id,
    location : env.GOOGLE_CLOUD_LOCATION || 'us-central1',
    pubSubSubscriptionName : env.GOOGLE_PUBSUB_SUBSCRIPTION_NAME || env.GCS_BUCKET_ENV || 'local-dev',
    gcsfuse : {
      rootDir : env.GCSFUSE_ROOT_DIR || '/etc/gcsfuse',

    },

    gcsDiskCache : {
      // size is in kilobytes
      maxSize : parseInt(env.GCS_DISK_CACHE_MAX_SIZE || 1000 * 1000),
      rootDir : env.GCS_DISK_CACHE_ROOT_DIR || '/etc/gcs-disk-cache',
      // age is in ms
      recheckAge : parseInt(env.GCS_DISK_CACHE_RECHECK_AGE || 1000 * 60 * 60),
      // extensions to be cached
      allowedExts : gcsDiskCacheExts || ['png', 'jpg', 'jpeg']
    },

    workflow : {
      type : 'gc-workflow',
      env : env.WORKFLOW_ENV,
      serviceAccountEmail : env.GOOGLE_CLOUD_WORKFLOW_SERVICE_ACCOUNT_EMAIL || env.GOOGLE_SERVICE_ACCOUNT_EMAIL || gcServiceAccount.client_email,
      maxConcurrentWorkflows : parseInt(env.GOOGLE_MAX_CONCURRENT_WORKFLOWS || 3),
      finWorkflowPath : '/fin/workflows/gc',
      timeoutMinutes : parseInt(env.GOOGLE_WORKFLOW_TIMEOUT_MINUTES || 30)
    }

  },

}
